SCClassLibrary/Common/Control/ServerStatus.sc

```diff
@@ -38,29 +38,93 @@ ServerStatusWatcher {
 		var doneOSCFunc, failOSCFunc;
+		var desiredClientID;
+
 		if(serverRunning.not) { ^this };
+
+		// flag true requests notification, false turns it off
 		notified = flag;
-		if(server.userSpecifiedClientID.not) {
-			doneOSCFunc = OSCFunc({|msg|
-				if(flag) { server.clientID = msg[2] };
-				failOSCFunc.free;
-			}, '/done', server.addr, argTemplate:['/notify', nil]).oneShot;
-
-			failOSCFunc = OSCFunc({|msg|
-				doneOSCFunc.free;
-				Error(
-					"Failed to register with server '%' for notifications: %\n"
-					"To recover, please reboot the server.".format(server.name, msg)).throw;
-			}, '/fail', server.addr, argTemplate:['/notify', nil, nil]).oneShot;
 
-		};
+		// set up oscfuncs for possible server responses, \done or \failed
+		doneOSCFunc = OSCFunc({|msg|
+			var newClientID = msg[2], newMaxLogins = msg[3];
+			failOSCFunc.free;
+
+			if(flag) {
+				// on registering scsynth sends back a free clientID and its maxLogins,
+				// which usually adjust the server object's settings:
+				this.prHandleClientLoginInfoFromServer(newClientID, newMaxLogins);
+			};
+
+		}, '/done', server.addr, argTemplate:['/notify', nil]).oneShot;
+
+		failOSCFunc = OSCFunc({|msg|
+
+			doneOSCFunc.free;
+			this.prHandleNotifyFailString(msg[2], msg);
+
+		}, '/fail', server.addr, argTemplate:['/notify', nil, nil]).oneShot;
+
+		// send the notify request - on or off, userSpecifiedClientID or not
+		desiredClientID = if(server.userSpecifiedClientID, { server.clientID }, {-1});
+		server.sendMsg("/notify", flag.binaryValue, desiredClientID);
 
 		if(flag){
-			"Receiving notification messages from server '%'\n".postf(server.name)
+			"Requested notification messages from server '%'\n".postf(server.name)
 		} {
 			"Switched off notification messages from server '%'\n".postf(server.name);
 		};
-		server.sendMsg("/notify", flag.binaryValue);
+	}
+
+	prHandleClientLoginInfoFromServer { |newClientID, newMaxLogins|
+		if (newMaxLogins.notNil) {
+			if (newMaxLogins != server.options.maxLogins) {
+				"%: scsynth has maxLogins % - adjusting my options accordingly.\n"
+				.postf(server, newMaxLogins);
+				server.options.maxLogins = newMaxLogins;
+			} {
+				"%: scsynth maxLogins % match with my options.\n".postf(server, newMaxLogins);
+			};
+		} {
+			"%: no maxLogins info from scsynth.\n".postf(server, newMaxLogins);
+		};
+
+		if (newClientID == server.clientID) {
+			"%: keeping clientID % as confirmed from scsynth.\n"
+			.postf(server, newClientID);
+			server.clientID = newClientID;
+		} {
+			if (server.userSpecifiedClientID.not) {
+				"%: setting clientID to %, as obtained from scsynth.\n"
+				.postf(server, newClientID);
+				server.clientID = newClientID;
+			} {
+				("% - userSpecifiedClientID % is not free!\n"
+					" Switching to free clientID obtained from scsynth: %.\n"
+					"If that is problematic, please set clientID by hand before booting.")
+				.format(server, server.clientID, newClientID).warn;
+				server.clientID = newClientID;
```

Comment:
all branches of this `if` block end with `server.clientID = newClientID`, could you please factor that out?

Reply:
(talking about lines 95, 100, 106)

Reply:
done.